core/node/da_clients/src/avail/client.rs

```diff
@@ -1,33 +1,132 @@
-use std::{fmt::Debug, sync::Arc};
-
+use anyhow::anyhow;
 use async_trait::async_trait;
+use bytes::Bytes;
 use jsonrpsee::ws_client::WsClientBuilder;
+use serde::{Deserialize, Serialize};
+use std::{fmt::Debug, sync::Arc};
 use subxt_signer::ExposeSecret;
-use zksync_config::configs::da_client::avail::{AvailConfig, AvailSecrets};
+use zksync_config::configs::da_client::avail::{AvailConfig, AvailSecrets, GasRelayAPIKey};
 use zksync_da_client::{
     types::{DAError, DispatchResponse, InclusionData},
     DataAvailabilityClient,
 };
+use zksync_types::{
+    ethabi::{self, Token},
+    web3::contract::{Tokenizable, Tokenize},
+    H256, U256,
+};
 
 use crate::avail::sdk::RawAvailClient;
 
 /// An implementation of the `DataAvailabilityClient` trait that interacts with the Avail network.
 #[derive(Debug, Clone)]
 pub struct AvailClient {
     config: AvailConfig,
-    sdk_client: Arc<RawAvailClient>,
+    sdk_client: Option<Arc<RawAvailClient>>,
+    api_client: Arc<reqwest::Client>,
+    gas_relay_api_key: Option<GasRelayAPIKey>,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct BridgeAPIResponse {
+    blob_root: Option<H256>,
+    bridge_root: Option<H256>,
+    data_root_index: Option<U256>,
+    data_root_proof: Option<Vec<H256>>,
+    leaf: Option<H256>,
+    leaf_index: Option<U256>,
+    leaf_proof: Option<Vec<H256>>,
+    range_hash: Option<H256>,
+    error: Option<String>,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct GasRelayAPISubmissionResponse {
+    submission_id: String,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct GasRelayAPIStatusResponse {
+    submission: GasRelayAPISubmission,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct GasRelayAPISubmission {
+    block_hash: Option<H256>,
+    extrinsic_index: Option<u64>,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct MerkleProofInput {
+    // proof of inclusion for the data root
+    data_root_proof: Vec<H256>,
+    // proof of inclusion of leaf within blob/bridge root
+    leaf_proof: Vec<H256>,
+    // abi.encodePacked(startBlock, endBlock) of header range commitment on vectorx
+    range_hash: H256,
+    // index of the data root in the commitment tree
+    data_root_index: U256,
+    // blob root to check proof against, or reconstruct the data root
+    blob_root: H256,
+    // bridge root to check proof against, or reconstruct the data root
+    bridge_root: H256,
+    // leaf being proven
+    leaf: H256,
+    // index of the leaf in the blob/bridge root tree
+    leaf_index: U256,
+}
+
+impl Tokenize for MerkleProofInput {
+    fn into_tokens(self) -> Vec<Token> {
```

Comment:
great job here! thanks for rewriting from alloy

Reply:
ty!